src/mlpack/methods/ann/layer/bilinear_interpolation_impl.hpp

```diff
@@ -0,0 +1,171 @@
+ *
+ * Implementation of the bilinear interpolation function as an individual layer.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license. You should have received a copy of the
+ * 3-clause BSD license along with mlpack. If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_BILINEAR_INTERPOLATION_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_BILINEAR_INTERPOLATION_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "bilinear_interpolation.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+
+template<typename InputDataType, typename OutputDataType>
+BilinearInterpolation<InputDataType, OutputDataType>::
+BilinearInterpolation():
+    inRowSize(0),
+    inColSize(0),
+    outRowSize(0),
+    outColSize(0),
+    depth(0)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+BilinearInterpolation<InputDataType, OutputDataType>::
+BilinearInterpolation(
+    const size_t inRowSize,
+    const size_t inColSize,
+    const size_t outRowSize,
+    const size_t outColSize,
+    const size_t depth):
+    inRowSize(inRowSize),
+    inColSize(inColSize),
+    outRowSize(outRowSize),
+    outColSize(outColSize),
+    depth(depth)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void BilinearInterpolation<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  if (output.is_empty())
+    output.set_size(outRowSize * outColSize * depth, 1);
+
+  assert(output.n_rows == outRowSize * outColSize * depth);
+  assert(output.n_cols == 1);
+
+  double scaleRow = (double)inRowSize / (double) outRowSize;
+  double scaleCol = (double)inColSize / (double) outColSize;
+
+  for (size_t k = 0; k < depth; k++)
+  {
+    for (size_t i = 0; i < outRowSize; i++)
+    {
+      for (size_t j = 0; j < outColSize; j++)
+      {
+        double rOrigin = std::floor(i * scaleRow);
+        double cOrigin = std::floor(j * scaleCol);
+
+        if (rOrigin > inRowSize - 2)
+          rOrigin = inRowSize - 2;
+        if (cOrigin > inColSize - 2)
+          cOrigin = inColSize - 2;
+
+        double deltaR = i * scaleRow - rOrigin;
+        double deltaC = j * scaleCol - cOrigin;
+        double coeff1 = (1 - deltaR) * (1 - deltaC);
+        double coeff2 = deltaR * (1 - deltaC);
+        double coeff3 = (1 - deltaR) * deltaC;
+        double coeff4 = deltaR * deltaC;
+
+        size_t ptr = k * inRowSize * inColSize + cOrigin * inColSize +
+                     rOrigin;
+
+        output(k * outRowSize * outColSize + j * outColSize + i) =
+            input(ptr) * coeff1 +
+            input(ptr + 1) * coeff2 +
+            input(ptr + inColSize) * coeff3 +
+            input(ptr + inColSize + 1) * coeff4;
+      }
+    }
+  }
```

Comment:
Is there any way that we could express this as a clearer or simpler Armadillo expression?  Here it looks like we are looping over each output element individually (and even accessing the output as a vectorized block of memory), but I think it could be a lot easier to understand what is going on if we can simplify the expression, and that may also allow the compiler to optimize.

One place to start might be with reinterpreting `output` as a different shape (should it be a cube?) so instead of accessing as `output(k * outRowSize * outColSize + j * outColSize + i)` you can do something more like `output(i, j, k)`.  You might consider doing that using an Armadillo alias: http://arma.sourceforge.net/docs.html#Cube (see the second advanced constructor).

Once you can do that, maybe you can put `coeff1` through `coeff4` in some vector `v` and write something a little more like this...

```
output(i, j, k) = input.slice(k).submat(rOrigin, cOrigin, rOrigin + 1, cOrigin + 1) % v;
```

That is just an idea although it is not perfect and would need a little bit of thinking.  Once you were able to do that, you might find that the entire triple for loop can be simplified and written as a single (or a few) Armadillo expressions.

Reply:
@rcurtin I think I can fixate the output on being a cube, but I think that way we'll have to lose the `OutputDataType` parameter. It is not a hard requirement anywhere as of now. Its needed for the O'Reilly test in the GAN PR.

Reply:
No need to change the output layer type, you can just make an alias like this, for instance:

```
arma::cube outputAsCube(output.memptr(), n_rows, n_cols, n_slices, false, true);
```

Reply:
@rcurtin Having given it some thought, I'm still a bit skeptical on how would this be achievable. I think, the triple `for` loop will have to be present, since the coefficients are calculated for all `i` and `j` and have to be applied on `k` slices as well. I came up with something like this (not sure of code correctness though):
```
if (output.is_empty())
    output.set_size(outRowSize * outColSize * depth, 1);

  assert(output.n_rows == outRowSize * outColSize * depth);
  assert(output.n_cols == 1);

  arma::cube InputAsCube(input.memptr(), inRowSize, inColSize, depth, false, true);
  arma::cube OutputAsCube(output.memptr(), outRowSize, outColSize, depth, false, true);

  double scaleRow = (double) inRowSize / (double) outRowSize;
  double scaleCol = (double) inColSize / (double) outColSize;

  arma::vec coeffs(4);
  for (size_t i = 0; i < outRowSize; i++)
  {
    for (size_t j = 0; j < outColSize; j++)
    {
      size_t rOrigin = (size_t) std::floor(i * scaleRow);
      size_t cOrigin = (size_t) std::floor(j * scaleCol);

      if (rOrigin > inRowSize - 2)
        rOrigin = inRowSize - 2;
      if (cOrigin > inColSize - 2)
        cOrigin = inColSize - 2;

      double deltaR = i * scaleRow - rOrigin;
      double deltaC = j * scaleCol - cOrigin;
      coeffs[0] = (1 - deltaR) * (1 - deltaC);
      coeffs[1] = deltaR * (1 - deltaC);
      coeffs[2] = (1 - deltaR) * deltaC;
      coeffs[3] = deltaR * deltaC;

      for (size_t k = 0; k < depth; k++)
      {
        OutputAsCube(i, j, k) = arma::accu(InputAsCube.slice(k).submat(
                rOrigin, cOrigin, rOrigin + 1,
                cOrigin + 1) % coeffs);
      }
    }
  }
```

Reply:
I definitely think that's an improvement.  You might be able to remove that last for loop (the one indexed by `k`) by using `.tube()` to access across all `depth` slices at once.  You might need to use `repmat()` on `coeffs` to make that work.

In any case, assuming that the result of the computation is the same as before, I think it is better and we should go with the newer version that you've made.  Thanks!

Reply:
I don't think that `repmat()` would be of much use either, since it will only return a matrix, while doing something like this (again not sure of the implementation):
```
 OutputAsCube.tube(i, j) = arma::accu(InputAsCube.tube(rOrigin, cOrigin, rOrigin + 1, cOrigin + 1) % repmat(coeffs, 1, depth));
```
we will run into a `sub_cube` on the input, and a `mat` on the coeffs (correct me if I'm wrong). But element wise multiplication wouldn't work here. What we need ideally is a `cube` version of the coeffs (I think), then the above should work.

Reply:
@rcurtin Also, doing above, I'll have to drop the `arma::accu`, and then do a `column-wise` and then `row-wise` addition to the element wise product matrix obtained. All in all, I don't think there'd be much to gain from making those changes. Should I just go with this implementation?